### ISSUE TITLE
Support `script setup` in Vue 3

### ISSUE THREAD

**Issue by arpadgabor** [storybookjs/storybook#16231]
**Is your feature request related to a problem? Please describe**

Vue 3 not long ago introduced official support for the `script setup` syntax that cleans up a lot of boilerplate in Vue components. Right now, with the latest beta, these components are not supported by `@storybook/vue`.

**Describe the solution you'd like**
Support for `script setup` components.

**Describe alternatives you've considered**
Write components normally with `defineComponent` but it becomes annoying.

**Are you able to assist to bring the feature to reality?**
I don't think so.

**Additional context**

![image](https://user-images.githubusercontent.com/9270494/135872947-9444ac98-f380-4b48-815f-bee7fa0fbce5.png)

Logged the imported component. It appears as it is not imported correctly.

<details>
<summary> The default component generated by Storybook, converted to `script setup`</summary>

```vue
<script setup lang="ts">
import { computed } from 'vue'

const emit = defineEmits(['click'])

const props = defineProps({
  label: {
    type: String,
    required: true,
  },
  primary: {
    type: Boolean,
    default: false,
  },
  size: {
    type: String,
    validator: function (value: string) {
      return ['small', 'medium', 'large'].indexOf(value) !== -1
    },
    required: true,
  },
  backgroundColor: {
    type: String,
    required: true,
  },
})

const classes = computed(() => ({
  'storybook-button': true,
  'storybook-button--primary': props.primary,
  'storybook-button--secondary': !props.primary,
  [`storybook-button--${props.size || 'medium'}`]: true,
}))

const style = computed(() => ({
  backgroundColor: props.backgroundColor,
}))

const onClick = () => {
  emit('click')
}
</script>

<template>
  <button type="button" :class="classes" :style="style" @click="onClick">
    {{ label }}
  </button>
</template>

<style>
.storybook-button {
  font-family: 'Nunito Sans', 'Helvetica Neue', Helvetica, Arial, sans-serif;
  font-weight: 700;
  border: 0;
  border-radius: 3em;
  cursor: pointer;
  display: inline-block;
  line-height: 1;
}
.storybook-button--primary {
  color: white;
  background-color: #1ea7fd;
}
.storybook-button--secondary {
  color: #333;
  background-color: transparent;
  box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
}
.storybook-button--small {
  font-size: 12px;
  padding: 10px 16px;
}
.storybook-button--medium {
  font-size: 14px;
  padding: 11px 20px;
}
.storybook-button--large {
  font-size: 16px;
  padding: 12px 24px;
}
</style>
```
</details>

**Comment by santiagoarizti** [comment]
note: as of v4.44.1 vue-docgen-api (and vue-styleguidist) support `<script setup>`
https://vue-styleguidist.github.io/docs/Documenting.html#setup-syntax
https://github.com/vue-styleguidist/vue-styleguidist/releases/tag/v4.44.0

**Comment by dLedda** [comment]
Hi, this issue is blocking us from migrating our project to script setup, which we are ready to do. Has any progress been made?

**Comment by mikestu** [comment]
More info:  It appears to also not work for <script setup> style components using Vue 2.7, where they backported the composition api from Vue 3.  We're stuck just not having Storybook controls for now.

**Comment by DoubleJ-G** [comment]
I also have this issue in Vue 2.7, although the components just don't work for me with errors/warning `[Vue warn]: Property or method "label" is not defined on the instance but referenced during render.` as if it's not working with script setup at all. 

@mikestu are you saying you have the components working but not the controls?

**Comment by CasperSocio** [comment]
I would like to see this extended to `<script setup lang="ts">` too.
My team has just made the transition to Vue3 with the goal of using Vite soon. Our styleguidist design system is now broken and I would very much like us to use Storybook instead. We're using Typescript along with script setup which is not properly supported in v6.5 and the workarounds results in a ton of cumbersome extra code that my team would like to stay clear of.

## Issues
The biggest issues using script setup in v6.5.

### Component description (Docs) must be manually added
The current Vue3 implementation gets the component description like this:

```vue
// BaseButton.vue

<script>
/** This text will show up as the component description */
export default {
  name: 'BaseButton',
}
</script>
```

Using script setup, we expected something like this to work

```vue
// BaseButton.vue

<script setup lang="ts">
/** This text should show up as the component description */

const props = defineProps<{
  // ...
}>()
</script>
```

but had to use this workaround instead

```typescript
// BaseButton.stories.ts

import BaseButton from './BaseButton.vue'

export default {
  title: 'Components/BaseButton',
  component: BaseButton,
  parameters: {
    docs: {
      description: {
        component: 'This text will show up as the component description'
      }
    }
  }
}
```

### Component source Doc Block not supporting composition
The source block should display only the template of a Vue component with the props applied. The current workaround

```typescript
export const CustomSource = Template.bind({});
CustomSource.parameters = {
  docs: {
    source: {
      code: 'Your code snippet goes here.',
      language: "yml",
      type: "auto",
    },
  },
}
```

is impractical and needs improvement. This might be solved by having Vue support on `.stories.(jsx|tsx)` files.

### Type support in stories
Better support for type suggestions in VS Code would be nice when modifying `.args`. This is well implemented with React, but not so much with Vue.

**Comment by CasperSocio** [comment]
> Still having this problem on vue `2.7.13` and storybook `6.5.12`

Storybook 7 addresses many of my complains, but is still lacking proper type support for component slots.

## Fixed by Storybook 7

- Doc-blocks directly above the meta object declaration is now displayed at the top of the docs section
- Doc-blocks directly above a prop definition is now visible in the props table
- Doc-blocks directly above story definitions are now displayed in the docs section

**Comment by matthew-dean** [comment]
@CasperSocio 

> Storybook 7 addresses many of my complains, but is still lacking proper type support for component slots.

Yes, I discovered the same. Have you filed an issue? There's a regression with Storybook 7 where `StoryFn` no longer types args to allow slots.

**Comment by shilman** [comment]
@CasperSocio @matthew-dean can one of you please file a separate issue for the typescript issue, preferably with a simple reproduction? i'd love to get that fixed before the 7.0 release. thank you!

**Comment by shilman** [comment]
@chakAs3 is this script setup issue something you can investigate for possible inclusion in 7.1?

**Comment by CasperSocio** [comment]
> @CasperSocio @matthew-dean can one of you please file a separate issue for the typescript issue, preferably with a simple reproduction? i'd love to get that fixed before the 7.0 release. thank you!

I'm on it!

**Comment by CasperSocio** [comment]
Made a new issue #21221 type support for Vue component slots.

**Comment by chakAs3** [comment]
> @chakAs3 is this script setup issue something you can investigate for possible inclusion in 7.1?

yes @shilman sure.

**Comment by chakAs3** [comment]
> Made a new issue #21221 type support for Vue component slots.

i ll look at it

**Comment by Walnussbaer** [comment]
Looking forward for a support of `script setup` sxntax. Esspecially some jsDoc at the top of the component that gets displayed in Storybook would be awesome :)

**Comment by chakAs3** [comment]
> Hi, this issue is blocking us from migrating our project to script setup, which we are ready to do. Has any progress been made?

I'm have already a PR addressing that, i believe it won't take me time to get it work for you, please reach me if you need some help for your migration.

**Comment by chakAs3** [comment]
> Looking forward for a support of `script setup` sxntax. Esspecially some jsDoc at the top of the component that gets displayed in Storybook would be awesome :)

can you explain me more your need, we ca collaborate and see how we can go faster inshallah.

**Comment by inkbeard** [comment]
I'm in the same boat here. Maybe I'm doing something incorrectly but basically, let's say there's a component called "AppButton.vue".

```
<script setup lang="ts">
  /** A custom button component for doing button-y things */
  // does some things
</script>

<template>
  <button>do a thing</button>
</template>
```

I can't find a way of including this description in the AppButton story. I've tried adding the `@description` tag, but it's not working. Using the component w/o the `setup` works fine. By that, I mean this works fine:

```
<script lang="ts">
  /** A custom button component for doing button-y things */

  export default {
    name: 'AppButton,
    ...
  }
</script>
...
```

**Comment by thedamon** [comment]
I see from testing, that this isn't fixed in Storybook 8 with or without vue-component-menat.

This is really, really needed. Script setup is the standard way of writing Vue components now and component descriptions are very important for docs. The autodocs approach is advertised as "the way" to be writing storybooks now, it should work with how we're encouraged to write Vue components!

**Comment by valentinpalkovic** [comment]
@huang-julien Can this be closed?
